### PR TITLE
Implement reverse playback and ping-pong loop in AnimationPlayer and NodeAnimation

### PR DESCRIPTION
**Pull request by TokageItLab**
For this proposal:
- [#2664](https://github.com/godotengine/godot-proposals/issues/2664) Add to support ping-pong loop in AnimationPlayer and playing backwards and ping-pong loop in NodeAnimation of AnimationTree

playmode

https://www.youtube.com/watch?v=mBOkTA4_HZE

**Note:**
This PR contains two PRs haven't merged as a prerequisite. One of them has been rejected in the past because its use was unclear, but this PR will require it. If you want to do the merge in stages, I recommend that you start with those PRs merges.

- [#41728](https://github.com/godotengine/godot/pull/41728) Implement reverse playback for animtree
- [#46346](https://github.com/godotengine/godot/pull/46346) Added pingpong built-in function

**Note2:**
One problem is that if the `TimeScale` is set to negative, it is not possible to tell when the animation is finished. This can be solved by using the `Backward` option in `NodeAnimation` itself.

For example, if you set
```
[Animation(forward)]->[TimeScale(-1)]->[OneShot]
```
this won't work as playing through backward animation (but if you want to temporarily reverse the animation, this is correct).

If you set
```
[Animation(backward)]->[TimeScale(1)]->[OneShot]
or
[Animation(forward)]->[OneShot]->[TimeScale(-1)]
````
these will work fine.

### CODE REVIEW

**Comment by fire** [comment]
When you are ready for review can you squash the merge commits?

**Comment by TokageItLab** [comment]
@fire Sure. I left a commit for other PRs to easily see, but I squashed it just now.

**Comment by fire** [comment]
Review from pouley said we wanted demos of more cases.

**Comment by TokageItLab** [comment]
@fire I uploaded a demo that works with some track types.

[ping-pong_demo.zip](https://github.com/godotengine/godot/files/6447173/ping-pong_demo.zip)

Some changes have also been made to the processing at both ends in order to support reverse playback. It may possibly conflict with #41168 and #41180.

As for RootMotion, ImmediateGeometry is not yet implemented in Vulkan, so #48568 is needed. There is a branch for development, including the PR, at https://github.com/TokageItLab/godot/tree/ping-pong-test-rootmotion

**Comment by reduz** [comment]
My feeling about this PR is that the implementation is very complex for something that is a very corner case. If it could be done in a way that it does not require modifying Animation and it could just be contained in AnimationPlayer and AnimationTree, it could be better.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
I see no other way to do this though, so I guess its ok.

File: scene/resources/audio_stream_sample.cpp [hunk 1]
```diff
@@ -650,7 +650,7 @@ void AudioStreamSample::_bind_methods() {
 
 	BIND_ENUM_CONSTANT(LOOP_DISABLED);
 	BIND_ENUM_CONSTANT(LOOP_FORWARD);
-	BIND_ENUM_CONSTANT(LOOP_PING_PONG);
+	BIND_ENUM_CONSTANT(LOOP_PINGPONG);
```
Comment (fire) [inline scene/resources/audio_stream_sample.cpp:653]: I'm curious is **PINGPONG** the more popular spelling compared to **PING_PONG**?

Comment (TokageItLab) [reply scene/resources/audio_stream_sample.cpp:653]: Since @Chaosus 's implementation of the math function is `pingpong()`, it is easy to see when mixed with other enumerated values, I decided to use `pingpong()`, `XXX_PINGPONG`, and `pingponged`.
However, since I am not so particular about it, I can consider using `XXX_PING_PONG` as the enumerated type and `ping_pong()` as the math function.

**Comment by TokageItLab** [comment]
@akien-mga Could you give me a review when you have a time?

**Comment by akien-mga** [comment]
Thanks!
